src/program.cpp

```diff
@@ -842,41 +867,75 @@ cl_build_status cvk_program::compile_source(const cvk_device* device) {
             return CL_BUILD_ERROR;
         }
     }
+#endif // CLSPV_ONLINE_COMPILER
 
-    // Compose clspv command-line
-    std::string cmd{gCLSPVPath};
+    bool binaryLoaded = false;
+    int status = -1;
     std::string spirv_file{tmp_folder + "/compiled.spv"};
+#ifdef CLSPV_ONLINE_COMPILER
+    if (build_from_il) {
+        build_options += " -o ";
+        build_options += spirv_file;
+        build_options += " ";
+        build_options += clspv_input_file;
+        temps.push(spirv_file);
+    }
 
-    temps.push(spirv_file);
-
+    cvk_info("About to compile \"%s\"", build_options.c_str());
+    // clspv is based on LLVM. LLVM options parsing is done using global
+    // variable that are not thread safe. Thus, we need to lock call to clspv in
+    // order to ensure a thread safe execution.
+    static std::mutex clspv_compile_lock;
+    clspv_compile_lock.lock();
     if (build_from_il) {
-        cmd += " -x ir ";
+        llvm::BumpPtrAllocator A;
+        llvm::StringSaver Saver(A);
+        llvm::SmallVector<const char*, 20> argv;
```

Comment:
Can you pass a plain old "const char *const argv" to clspv::Compile?

Reply:
clspv::Compile expects an already tokenized const char *const argv[]. https://github.com/google/clspv/blob/14378ce9e67d6727ce24578ef98811b6395fc5df/lib/Compiler.cpp#L942

I wanted to try to be consistent with how the command line options are built and tokenized to try to ensure a uniform behavior when running in different configurations so I followed the same approach used by clspv::CompileFromSourceString https://github.com/google/clspv/blob/14378ce9e67d6727ce24578ef98811b6395fc5df/lib/Compiler.cpp#L1075